File: singleton/java/LazySingleton.java
// 懒汉单例
// 节约内存，但线程不安全
// 修改方式是在 getInstance 前加上 synchronized 关键字，但是这样会造成很大的性能开销
public class LazySingleton {

  private LazySingleton() {}

  private static LazySingleton instance;

  public static synchronized LazySingleton getInstance() {
    if (instance == null) {
      instance = new LazySingleton();
    }
    return instance;
  }

  public void check() {
    System.out.println("the instance has been created");
  }

  public static void main(String[] args) {
    getInstance().check();
  }
}

File: singleton/python/LazySingleton.py
#!/usr/bin/env python
# coding:utf8

class LazySingleton(object):
  __instance = None

  @classmethod
  def getInstance(cls):
    if LazySingleton.__instance is None:
      LazySingleton.__instance = object.__new__(cls)
    return LazySingleton.__instance

  def check(self):
    print('the instance has been created')

  def __init__(self):
    raise 'createError'

if __name__ == '__main__':
  LazySingleton.getInstance().check()

# 无法通过 m n o 的原因是单例的获取有多种方式。